File: cpp/src/alerts.cpp
#include <rtxmon/alerts.hpp>

#include <stdexcept>

namespace rtxmon {

AlertEvaluator::AlertEvaluator(AlertOptions options)
    : options_(options)
{
    if (options_.hysteresis_c < 0) {
        throw std::invalid_argument("alert hysteresis must not be negative");
    }
    if (options_.hysteresis_c > options_.threshold_c) {
        throw std::invalid_argument("alert hysteresis must not exceed the threshold");
    }
}

std::optional<TelemetryEventKind> AlertEvaluator::observe(std::int32_t temperature_c) noexcept
{
    if (!alarmed_ && temperature_c >= options_.threshold_c) {
        alarmed_ = true;
        return TelemetryEventKind::alert_raised;
    }

    const auto clear_temperature = options_.threshold_c - options_.hysteresis_c;
    const bool dropped_below_threshold = options_.hysteresis_c == 0
        ? temperature_c < clear_temperature
        : temperature_c <= clear_temperature;
    if (alarmed_ && dropped_below_threshold) {
        alarmed_ = false;
        return TelemetryEventKind::alert_cleared;
    }

    return std::nullopt;
}

bool AlertEvaluator::alarmed() const noexcept
{
    return alarmed_;
}

const AlertOptions &AlertEvaluator::options() const noexcept
{
    return options_;
}

} // namespace rtxmon

File: csharp/RtxMonitor.Managed/Alerts.cs
namespace RtxMonitor.Managed;

public sealed record AlertOptions(int ThresholdC, int HysteresisC = 0);

// Pure state machine: turns a stream of die temperatures into AlertRaised /
// AlertCleared transitions. Holds no session, thread, or clock of its own,
// so it stays testable without a GPU and independent of ResilientSampler's
// reconnect/backoff policy.
public sealed class AlertEvaluator
{
    private readonly AlertOptions options;
    private bool alarmed;

    public AlertEvaluator(AlertOptions options)
    {
        if (options.HysteresisC < 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(options),
                "A histerese do alerta não pode ser negativa.");
        }
        if (options.HysteresisC > options.ThresholdC)
        {
            throw new ArgumentOutOfRangeException(
                nameof(options),
                "A histerese do alerta não pode exceder o limiar.");
        }

        this.options = options;
    }

    public bool Alarmed => alarmed;

    public AlertOptions Options => options;

    public TelemetryEventKind? Observe(int temperatureC)
    {
        if (!alarmed && temperatureC >= options.ThresholdC)
        {
            alarmed = true;
            return TelemetryEventKind.AlertRaised;
        }

        int clearTemperature = options.ThresholdC - options.HysteresisC;
        bool droppedBelowThreshold = options.HysteresisC == 0
            ? temperatureC < clearTemperature
            : temperatureC <= clearTemperature;
        if (alarmed && droppedBelowThreshold)
        {
            alarmed = false;
            return TelemetryEventKind.AlertCleared;
        }

        return null;
    }
}
